From 17f4a06be88ed9c950e562aa72755c7a493b8ae9 Mon Sep 17 00:00:00 2001
From: Niall Pemberton <niallp@apache.org>
Date: Tue, 6 Oct 2009 15:05:02 +0000
Subject: [PATCH] BEANUTILS-345 BeanUtilsBean.setProperty does not handle some
 kind of nested properties - thanks to Simone Riccucci for the patch

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/beanutils/trunk@822308 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/beanutils/BeanUtilsBean.java      |   2 +
 .../beanutils/bugs/Jira345TestCase.java       | 119 ++++++++++++++++++
 2 files changed, 121 insertions(+)
 create mode 100644 src/test/org/apache/commons/beanutils/bugs/Jira345TestCase.java

diff --git a/src/java/org/apache/commons/beanutils/BeanUtilsBean.java b/src/java/org/apache/commons/beanutils/BeanUtilsBean.java
index a9ac122df..0fffebe6e 100644
--- a/src/java/org/apache/commons/beanutils/BeanUtilsBean.java
+++ b/src/java/org/apache/commons/beanutils/BeanUtilsBean.java
@@ -927,6 +927,8 @@ public void setProperty(Object bean, String name, Object value)
             type = dynaProperty.getType();
         } else if (target instanceof Map) {
             type = Object.class;
+        } else if (target.getClass().isArray() && index >= 0) {
+            type = Array.get(target, index).getClass();
         } else {
             PropertyDescriptor descriptor = null;
             try {
diff --git a/src/test/org/apache/commons/beanutils/bugs/Jira345TestCase.java b/src/test/org/apache/commons/beanutils/bugs/Jira345TestCase.java
new file mode 100644
index 000000000..ad8e0b930
--- /dev/null
+++ b/src/test/org/apache/commons/beanutils/bugs/Jira345TestCase.java
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.beanutils.bugs;
+
+import junit.framework.Test;
+import junit.framework.TestCase;
+import junit.framework.TestSuite;
+
+import org.apache.commons.beanutils.BeanUtils;
+
+/**
+ * See https://issues.apache.org/jira/browse/BEANUTILS-345
+ * <p />
+ *
+ * @version $Revision$ $Date$
+ */
+public class Jira345TestCase extends TestCase {
+
+    /**
+     * Create a test case with the specified name.
+     *
+     * @param name The name of the test
+     */
+    public Jira345TestCase(String name) {
+        super(name);
+    }
+
+    /**
+     * Run the Test.
+     *
+     * @param args Arguments
+     */
+    public static void main(String[] args) {
+        junit.textui.TestRunner.run(suite());
+    }
+
+    /**
+     * Create a test suite for this test.
+     *
+     * @return a test suite
+     */
+    public static Test suite() {
+        return (new TestSuite(Jira345TestCase.class));
+    }
+
+    /**
+     * Set up.
+     *
+     * @throws java.lang.Exception
+     */
+    protected void setUp() throws Exception {
+        super.setUp();
+    }
+
+    /**
+     * Tear Down.
+     *
+     * @throws java.lang.Exception
+     */
+    protected void tearDown() throws Exception {
+        super.tearDown();
+    }
+
+    /**
+     * Test {@link BeanUtils} setProperty() with 2D array.
+     */
+    public void testBeanUtilsSetProperty_2DArray() throws Exception{
+        MyBean myBean = new MyBean();
+        BeanUtils.setProperty(myBean, "matr[0][0]","Sample");
+        assertEquals("Sample", myBean.getMatr()[0][0]);
+    }
+
+    /**
+     * Test {@link BeanUtils} setProperty() with 3D array.
+     */
+    public void testBeanUtilsSetProperty_3DArray() throws Exception{
+        MyBean myBean = new MyBean();
+        BeanUtils.setProperty(myBean, "matr3D[0][0][0]","Sample");
+        assertEquals("Sample", myBean.getMatr3D()[0][0][0]);
+    }
+
+    /** Example Bean */
+    public static class MyBean {
+
+        private String[][] matr = new String[][]{{"1","2"},{"3","4"}};
+
+        private String[][][] matr3D = new String[][][] {
+                {{"11","12"}, {"13","14"}},
+                {{"21","22"}, {"23","24"}},
+        };
+        
+        public String[][] getMatr() {
+            return matr;
+        }
+        public void setMatr(String[][] matr) {
+            this.matr = matr;
+        }
+        public String[][][] getMatr3D() {
+            return matr3D;
+        }
+        public void setMatr3D(String[][][] matr3D) {
+            this.matr3D = matr3D;
+        }
+    }
+}
